sqlmesh/core/snapshot/evaluator.py

```diff
@@ -88,15 +88,30 @@ class SnapshotEvaluator:
     does not directly communicate with the underlying execution engine.
 
     Args:
-        adapter: The adapter that interfaces with the execution engine.
+        adapters: A single EngineAdapter or a dictionary of EngineAdapters where
+            the key is the gateway name. When a dictionary is provided, its first
+            item is treated as the default adapter and used for the virtual layer.
         ddl_concurrent_tasks: The number of concurrent tasks used for DDL
             operations (table / view creation, deletion, etc). Default: 1.
     """
 
-    def __init__(self, adapter: EngineAdapter, ddl_concurrent_tasks: int = 1):
-        self.adapter = adapter
+    def __init__(
+        self,
+        adapters: EngineAdapter | t.Dict[str, EngineAdapter],
+        ddl_concurrent_tasks: int = 1,
+    ):
+        self.adapters = adapters if isinstance(adapters, t.Dict) else {"": adapters}
+        self.adapter = next(iter(self.adapters.values()))
```

Comment:
I'd made a similar comment previously, but it [seems like](https://github.com/TobikoData/sqlmesh/pull/3394#discussion_r1854337841) the first one is always gonna be the primary adapter because that's set in the [dictionary first](https://github.com/TobikoData/sqlmesh/pull/3394/files#diff-02c2689e9e8516b4f905e12fa4dac9b0fff86efae7964c29fb9337055f6f3d94R377-R379).

Reply:
one alternative would be to pass again the default_adapter? or maybe the `gateway` or `default_gateway` should be passed as a string to the evaluator when initialising?

Reply:
Went with the latter and using as a fallback the first adapter if the gateway isn't passed on to the evaluator